### ISSUE TITLE
JS chunk filenames & caching

### ISSUE THREAD

**Issue by TomiTakussaari** [gatsbyjs/gatsby#2538]
Hi

So I just spend some time investigating why our site started to break with "Uncaught TypeError: Cannot read property 'call' of undefined" when webpack is bootstrapping itself.

It turned out that problem was that JS file name generated by gatsby/webpack do not seem to (always?) change when content is modified, and that combined with the fact that I had caching rule that basically cached all JS files for one week, somehow resulted in application that tries to use both old and new code, which usually failed.

Clearing caches in browsers and CDN solved the problem for now atleast.

But is this by design or somekind of configuration error / bug ? 

I expected that filename would change if content changes, but that does not seem to be so, or then something not obvious to me is going on.

There does not seem to be any documentation about caching in Gatsby site, so perhaps we need to add that ? I would be glad to contribute, once I understand what is going on :)

Using gatsby 1.9.73.

**Comment by KyleAMathews** [comment]
Hmm yeah — I've meant to add a caching docs page but clearly forgot so would appreciate one from you.

But yeah, webpack 1 basically makes it impossible to make long-term caching a thing. Webpack 3 is supposed to so fingers crossed. So for now, don't set a cache on your JS.

Sorry for the bother!

**Comment by monsieurnebo** [comment]
After each new deployment, I'm facing a blank page caused by the following error:

```
Uncaught TypeError: Cannot read property 'call' of undefined
    at t (bootstrap 31e8a463e48ea4735095:52)
    at index.json?9cdc:9
    at loader.js:96
    at loader.js:81
    at index.json?9cdc:9
    at c (patch.js:19)
    at patch.js:40
    at window.webpackJsonp (bootstrap 31e8a463e48ea4735095:20)
    at path---index-a0e39f21c11f6a62c5ab.js:1
```

A simple page refresh will fix it, but it's not a viable solution. It seems to be related to this issue so... Have you any workaround awaiting the Webpack 3 integration?

**Comment by charlesgiroux** [comment]
As a workaround I am using the following in gatsby-node.js:

```javascript
exports.modifyWebpackConfig = ({ config, stage }) => {
  const timestamp = Date.now()
  switch (stage) {
    case 'build-javascript':
      config.merge({
        output: {
          filename: `[name]-${timestamp}-[chunkhash].js`,
          chunkFilename: `[name]-${timestamp}-[chunkhash].js`
        }
      })

      break
  }
  return config
}
```

This allows the CDN to cache JS files until I push a new build.

**Comment by KyleAMathews** [comment]
If you haven't seen it, I wrote up a new page recently on how to cache Gatsby's assets https://www.gatsbyjs.org/docs/caching/

tl;dr set must-revalidate for JS files for now.

**Comment by charlesgiroux** [comment]
Yeah I saw it, but I would rather have the files cached for some time than not at all.

**Comment by BlakeTurner** [comment]
@KyleAMathews knowing the caching limitations of Webpack 1 in Gatsby 1, @charlesgiroux's solution is probably worth a mention in the docs. It cache busts every file in every build, but also allows the browser to cache the files on more than one load of the site or app. I think it's an ideal mix.

**Comment by KyleAMathews** [comment]
@BlakeTurner the problem with that method is that then users have to redownload the script files every time instead of just validating that the script hasn't changed. Ideally they don't have to validate but there's no reason to make people download the same file every time.

**Comment by charlesgiroux** [comment]
@BlakeTurner I think @KyleAMathews is right. "cache-control: public, max-age=0, must-revalidate" is more appropriate for most apps/sites. 

I only use my solution for 1 site where the JS files should not change more than once per 6 months, if it changes then I need to make sure the cache gets completely busted.

**Comment by BlakeTurner** [comment]
@KyleAMathews Definitely see where you're coming from. I think my problem is that I don't understand _exactly_ why a new deployment + cached JS is causing the `Cannot read property 'call' of undefined` error. Is there somewhere in the Gatsby docs I can get into the nuts and bolts of this? Or is my best bet the source? Basically I'm looking for the steps that occur from page request to React loaded in the browser - how the different JS files are loaded async, how the app knows which version to use, etc.

Thanks!

**Comment by KyleAMathews** [comment]
It's webpack trying to call a function on code that doesn't exist. When you load an HTML file from Gatsby, includes mappings from module ids to bundles. If the mapping changes but the bundle name doesn't, webpack will load the cached bundle with now outdated modules and try to call code there and the code won't be there hence "undefined"

**Comment by bigfanjs** [comment]
@KyleAMathews Thank you very much! I appreciate your detailed explanations of how it all happens. I am not sure if it's possible but I would appreciate it even more if you suggest a way to display something useful to users when such crash(**Uncaught TypeError: Cannot read property 'call' of undefined"**) happens in the bootstrap phase.

**Comment by garethweaver** [comment]
@charlesgiroux new config edit here:

```
exports.onCreateWebpackConfig = ({ stage, getConfig, actions }) => {
  if (stage === 'build-javascript') {
    const timestamp = Date.now()
    let config = getConfig()
    config.output.filename = `[name]-${timestamp}-[chunkhash].js`
    config.output.chunkFilename = `[name]-${timestamp}-[chunkhash].js`
    actions.replaceWebpackConfig(config)
  }
}
```

**Comment by NickScottKortical** [comment]
If anyone is using gatsby with nginx and kubernetes pods/canary deployments, they may need to ensure they enable sticky sessions within their nginx config. This solved the issue for me so that the js requests would go to the same pod that served the html content.

``` 
    nginx.ingress.kubernetes.io/affinity: "cookie"
    nginx.ingress.kubernetes.io/session-cookie-name: "route"
    nginx.ingress.kubernetes.io/session-cookie-hash: "sha1"
```
